[OpenMP] Taskgraph Clang 'record and replay' frontend support

jtb20 opened the pull request:
This is a version of the `ompx taskgraph` support posted in [PR66919](https://github.com/llvm/llvm-project/pull/66919), adapted to the official OpenMP 6.0 spelling of `omp taskgraph`, and with the `ompx` extension parts removed.

I've merged the changes to the current mainline, tidied up the patch series and adjusted tests a little, but other than that I've not done much testing, so support should still be considered experimental. See also https://github.com/llvm/llvm-project/issues/130926.

jtb20 committed:
[OpenMP] Taskgraph frontend support

This is a version of the 'ompx taskgraph' support posted in PR66919,
adapted to the official OpenMP 6.0 spelling of 'omp taskgraph', and with
the 'ompx' extension parts removed.

Co-authored-by: Adrian Munera <adrian.munera@bsc.es>
Co-authored-by: Jose M Monsalve Diaz <JoseM.MonsalveDiaz@amd.com>

jtb20 committed:
[OpenMP] New/derived taskgraph tests

This patch adds new tests for 'omp taskgraph' functionality, but (unlike
the patches posted in PR66919) leave the existing tests using the internal
runtime API for record and replay as-is.

I have changed the 'print_dot' tests to use FileCheck instead of their
own internal checking, though.

Co-authored-by: Adrian Munera <adrian.munera@bsc.es>

DeinAlptraum reviewed (commented):
Re Libclang-Python bindings: LGTM

alexey-bataev reviewed (commented):
clang unit tests (parsing, sema, codegen) are required.
Also, need to update OpenMP.rst and ReleseNotes docs

DeinAlptraum commented on clang/bindings/python/clang/cindex.py:
```diff
@@ -1446,6 +1446,9 @@ def is_unexposed(self):
     # OpenMP stripe directive.
     OMP_STRIPE_DIRECTIVE = 310
 
+    # OpenMP taskgraph directive.
+    OMP_TASKGRAPH_DIRECTIVE = 311
+
```
Good point,  completely forgot about that
Reference: https://github.com/llvm/llvm-project/pull/139293

jtb20 replied:
I've changed the number to 312 (but I don't think the other patch is committed yet, so there's a hole in the numbering for now).

jtb20 committed:
[OpenMP] New Clang tests for 'taskgraph' directive

This patch adds two simple tests for parsing/serialization/deserialization
and codegen for the OpenMP 'taskgraph' directive (for the current
record-and-replay based implementation).

jtb20 commented on clang/lib/CodeGen/CGOpenMPRuntime.cpp:
```diff
@@ -5954,6 +5976,48 @@ void CGOpenMPRuntime::emitTaskwaitCall(CodeGenFunction &CGF, SourceLocation Loc,
+                                                  &TaskgraphRegion);
+  llvm::Function *FnT = OutlinedCGF.GenerateCapturedStmtFunction(*CS);
+
+  SmallVector<llvm::Value *, 6> Args{
```
Done.

jtb20 commented:
> clang unit tests (parsing, sema, codegen) are required. Also, need to update OpenMP.rst and ReleseNotes docs

I added a couple of tests, but forgot about the docs. I'll push again in a bit...

jtb20 commented:
@alexey-bataev actually I'm not sure these patches bring the 'taskgraph' implementation to a level that is ready to announce as a supported feature in the documentation yet -- what did you have in mind?

alexey-bataev commented:
> @alexey-bataev actually I'm not sure these patches bring the 'taskgraph' implementation to a level that is ready to announce as a supported feature in the documentation yet -- what did you have in mind?

Then mark it as 'partially supported' or "in progress"

jtb20 committed:
[OpenMP] Mention partial taskgraph support in Clang release notes

NFC.

jtb20 commented:
This rebased version mentions partial taskgraph support in `clang/docs/ReleaseNotes.rst`. There's already a table entry regarding record-and-replay taskgraph support's in-progress status in `clang/docs/OpenMPSupport.rst`.

alexey-bataev reviewed (commented):
Pl
